File: src/main/java/com/pandus/leetcode/solutions/google/DailyTemperatures.java
package com.pandus.leetcode.solutions.google;

import java.util.ArrayDeque;

public class DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        ArrayDeque<Integer> st = new ArrayDeque<>();
        int[] ans = new int[temperatures.length];
        for (int i = 0; i < temperatures.length; i++) {
            int c = temperatures[i];
            while (!st.isEmpty() && temperatures[st.peek()] < c) {
                int idx = st.pop();
                ans[idx] = i - idx;
            }
            st.push(i);
        }
        return ans;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/daily/DailyTemperatures.kt
package com.pandus.leetcode.solutions.daily

// Reference: https://leetcode.com/problems/daily-temperatures
class DailyTemperatures {
    fun dailyTemperatures(temperatures: IntArray): IntArray {
        val result = IntArray(temperatures.size)
        val stack = ArrayDeque<Int>()

        for ((index, temperature) in temperatures.withIndex()) {
            while (stack.isNotEmpty() && temperatures[stack.last()] < temperature) {
                val day = stack.removeLast()
                result[day] = index - day
            }
            stack.add(index)
        }

        return result
    }
}
